jerry-core/ecma/operations/ecma-promise-object.c

```diff
@@ -0,0 +1,711 @@
+ */
+inline void __attr_always_inline___
+ecma_promise_set_state (ecma_object_t *obj_p, /**< points to promise object */
+                        uint8_t state) /**< the state */
+{
+  JERRY_ASSERT (ecma_is_promise (obj_p));
+
+  ((ecma_promise_object_t *) obj_p)->state = state;
+} /* ecma_promise_set_state */
+
+/**
+ * Get the bool value of alreadyResolved.
+ *
+ * @return bool value of alreadyResolved.
+ */
+static bool
+ecma_get_already_resolved_bool_value (ecma_value_t already_resolved) /**< the alreadyResolved */
+{
+  JERRY_ASSERT (ecma_is_value_object (already_resolved));
+
+  ecma_object_t *already_resolved_p = ecma_get_object_from_value (already_resolved);
+  ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) already_resolved_p;
+
+  JERRY_ASSERT (ext_object_p->u.class_prop.class_id == LIT_MAGIC_STRING_BOOLEAN_UL);
+
+  return ext_object_p->u.class_prop.u.value == ecma_make_boolean_value (true);
+} /* ecma_get_already_resolved_bool_value */
+
+/**
+ * Set the value of alreadyResolved.
+ */
+static void
+ecma_set_already_resolved_value (ecma_value_t already_resolved, /**< the alreadyResolved */
+                                 bool value) /**< the value to set */
+{
+  JERRY_ASSERT (ecma_is_value_object (already_resolved));
+
+  ecma_object_t *already_resolved_p = ecma_get_object_from_value (already_resolved);
+  ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) already_resolved_p;
+
+  JERRY_ASSERT (ext_object_p->u.class_prop.class_id == LIT_MAGIC_STRING_BOOLEAN_UL);
+
+  ext_object_p->u.class_prop.u.value = ecma_make_boolean_value (value);
+} /* ecma_set_already_resolved_value */
+
+/**
+ * Take a collection of Reactions and enqueue a new PromiseReactionJob for each Reaction.
+ *
+ * See also: ES2015 25.4.1.8
+ */
+static void
+ecma_promise_trigger_reactions (ecma_collection_header_t *reactions, /**< lists of reactions */
+                                ecma_value_t value) /**< value for resolve or reject */
+{
+  ecma_collection_iterator_t iter;
+  ecma_collection_iterator_init (&iter, reactions);
+
+  while (ecma_collection_iterator_next (&iter))
+  {
+    ecma_enqueue_promise_reaction_job (*iter.current_value_p, value);
+  }
+
+  ecma_free_values_collection (reactions, false);
+} /* ecma_promise_trigger_reactions */
+
+/**
+ * Reject a Promise with a reason.
+ *
+ * See also: ES2015 25.4.1.7
+ */
+static void
+reject_promise (ecma_value_t promise, /**< promise */
+                ecma_value_t reason) /**< reason for reject */
+{
+  ecma_object_t *obj_p = ecma_get_object_from_value (promise);
+
+  JERRY_ASSERT (ecma_promise_get_state (obj_p) == ECMA_PROMISE_STATE_PENDING);
+
+  ecma_promise_set_result (obj_p, ecma_copy_value_if_not_object (reason));
+
+  ecma_promise_object_t *promise_p = (ecma_promise_object_t *) obj_p;
+  ecma_promise_trigger_reactions (promise_p->reject_reactions, reason);
+  promise_p->reject_reactions = ecma_new_values_collection (NULL, 0, false);
```

Comment:
Yes, it can. But if so, when we add reactions to the list, we have to create a collection when the value is NULL